### PR TITLE
feat(tui): copy selected tool output

### LINKED ISSUE #133: TUI tools: copy selected tool output
## Goal
Allow copying an individual tool result without copying the whole transcript.

## Acceptance criteria
- User can copy selected/focused tool output.
- Copy failures are rendered in the transcript.
- Large output behavior is clear.
- Add tests for successful and empty/error states.

### PR DESCRIPTION
**Pull request by lsm**
Closes #133

## Summary
- Track and render the selected tool row in TUI state/tool panel
- Add selected tool output copying via Ctrl+X and `/copy tool`
- Render transcript success, empty-output, and large-output copy messages
- Add unit coverage for successful, empty, large, command, and selection/render states

## Tests
- `zig build test-unit-makai-cli`
- `zig build test-unit-core && zig build test-unit-transport && zig build test-unit-protocol && zig build test-unit-providers && zig build test-unit-makai-cli && zig build test-unit-agent`
- `../scripts/check-zig-patterns.sh`

## Known test issue
- `zig build test` currently fails in `test-unit-utils` with existing Codex OAuth mock output: `Codex OAuth error: invalid_grant - bad code`.

### CODE REVIEW

File: zig/src/tui/state.zig [hunk 1]
```diff
@@ -1399,6 +1412,7 @@ pub const AppState = struct {
             return tool;
         }
         try self.tools.append(self.allocator, try ToolEntry.init(self.allocator, id, name, label, args_json, status));
+        self.selected_tool_index = self.tools.items.len - 1;
```
Comment (lsm) [inline zig/src/tui/state.zig]: **P1 — selection clobbering.** Setting `selected_tool_index` to the newest tool on every append runs on the live event path (all four `applyEvent` tool branches call `upsertTool` at state.zig:1030/1033/1039/1045), and tool events arrive incrementally during a turn. So if the user focuses the tools pane, navigates **up** to an older tool to copy it, and another tool streams in mid-turn, this line silently overwrites their selection: the highlight jumps to the newest tool and `/copy tool` / `Ctrl+X` copies the wrong (newest) tool instead of the one the user navigated to. There is no "user is navigating" guard (unlike the transcript pane's `manual_transcript_paging`).

This directly undermines the headline use case of #133 (copy an *individual* tool result, including older ones).

Suggested fix: move the "default to latest" behavior into a fallback in `selectedToolOutputText()` (return last tool's output when `selected_tool_index == null`) and **remove** this mutation from `upsertTool`. That preserves "no explicit selection → latest tool" without ever clobbering an explicit selection. Alternatively, gate this line behind a `manual_tool_paging`-style flag set by `moveSelection(.tools)`.

The new test `"AppState tracks selected tool output"` doesn't catch this because both appends happen *before* the `moveSelection(-1)` — it never exercises navigate-then-append, which is the real race.

File: zig/src/tui/app.zig [hunk 2]
```diff
@@ -1311,6 +1312,19 @@ pub const App = struct {
         self.state.appendTranscript(.system, "copied preview to clipboard") catch {};
     }
 
+    fn copySelectedToolOutput(self: *App) void {
+        const text = self.state.selectedToolOutputText() orelse {
+            self.state.appendTranscript(.system, "selected tool has no output to copy") catch {};
+            return;
+        };
+        self.stageClipboard(text);
```
Comment (lsm) [inline zig/src/tui/app.zig:1324]: **P2 — clipboard write failures are silently swallowed.** `stageClipboard` only dupes the text; the actual write happens later in `flushClipboard`, which does `_ = ctx.setClipboard(text) catch {};` (app.zig:1278). So if the OSC 52 clipboard write fails, the transcript still says "copied tool output to clipboard" — a false success — and nothing is rendered for the failure.

The acceptance criteria list "Copy failures are rendered in the transcript" and a "Clipboard failure renders transcript error" test, but no such test exists and the write-error path can't surface through the current `stageClipboard` → `flushClipboard` split (flush is void and runs later, possibly without a `ctx`).

Two acceptable resolutions:
1. Surface the failure: have `flushClipboard` return/record the error and render an `[error]` transcript entry on failure (would also fix the same latent gap in `copyLastAssistant` / `copyTranscript` / `copyPreview`).
2. If OSC 52 failures aren't reliably detectable (often true), keep current behavior but update the PR description/acceptance to state that only the *no-output* failure path is reported, so the criterion isn't claimed but unmet.

The empty/no-selection path *is* correctly rendered ("selected tool has no output to copy") and tested.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (ZhipuAI)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** ZhipuAI

**Recommendation: REQUEST_CHANGES** (own-PR, so posting as COMMENT — GitHub rejects REQUEST_CHANGES from the author. Please treat this as a changes-requested verdict.)

The feature works in the common (idle) case and the unit tests pass locally (`zig build test-unit-tui`, exit 0) and the pattern guardrail is clean. But two issues should be addressed before merge: one is a correctness defect in the headline use case, the other means the PR's own tests don't run in CI.

### What's good
- Clean wiring of the new action: `commands.zig` (`/copy tool`), `app.zig` routing (`.copy_tool`), and `Ctrl+X` keybinding. Both copy paths flush the clipboard (`flushClipboard` at app.zig:2046 for the command path; inline at app.zig:1764 for the keybinding).
- Empty/no-selection path renders a clear transcript message and is tested.
- Large-output threshold (256 KiB) with a distinct message, tested with a >256 KiB payload.
- `Ctrl+X` has no conflicts (single handler at app.zig:1762; the `'x'` matches elsewhere are unrelated test data).

### P1 — `upsertTool` clobbers explicit user selection during streaming (state.zig:1415)
`self.selected_tool_index = self.tools.items.len - 1;` runs on **every** new tool append — i.e. on the live event path (all four `applyEvent` tool handlers call `upsertTool`). Tool events arrive incrementally during a turn, so if the user focuses the tools pane, navigates **up** to an older tool to copy it, and another tool streams in mid-turn, their selection is silently overwritten to the newest tool. The render highlight jumps and `/copy tool` / `Ctrl+X` copies the wrong tool. No "user is navigating" guard exists (contrast the transcript pane's `manual_transcript_paging`). This directly undermines #133's purpose (copy an *individual*, possibly older, tool). See anchored comment with a suggested fix (fallback in `selectedToolOutputText` + remove the mutation, or gate behind a manual-nav flag).

### P1 — the PR's new tests are not exercised by CI
`.github/workflows/ci.yml:50` matrix is `[core, transport, protocol, providers, utils, makai-cli, agent-types, agent-loop, agent-mod, agent-bridge, agent-unit, agent-chain]` — there is no `tui` group, and CI never runs `zig build test` (only `test-unit-${matrix.group}`). All four tests added by this PR live in the TUI test binaries (`tui_state_test`, `tui_commands_test`, `tui_app_test`) that are only reached via `test-unit-tui` / the `test` step. So the new tests do not run in CI, contrary to the standing instruction that updated tests must pass in CI. (Pre-existing gap, but it blocks this PR's acceptance.) Recommended fix: add `tui` to the `matrix.group` list (one line). Side note: the PR body lists the test command as `test-unit-makai-cli`, which does **not** cover these files — the correct group is `test-unit-tui`.

### P2 — clipboard write failures are not rendered (acceptance gap)
The acceptance criteria call for "Copy failures rendered in the transcript" and a "Clipboard failure renders transcript error" test. The empty/no-selection path is handled and tested, but the actual clipboard write goes through `flushClipboard` → `ctx.setClipboard(text) catch {}` (app.zig:1278), which swallows errors silently; on failure the transcript still claims "copied tool output to clipboard". No test covers this. Pre-existing pattern shared by all copy commands. Either surface flush errors into an `[error]` transcript entry, or relax the criterion and document that only the no-output failure path is reported (OSC 52 can't reliably error). See anchored comment.

### P3 — `display_preview` preference bypasses large-output detection for artifact tools (informational)
`selectedToolOutputText()` (state.zig:679) returns `display_preview` first when set. `display_preview` is only populated for artifact-producing tools and is capped at `artifact_display_preview_read_limit` = 256 KiB (app.zig:40). So for an artifact tool with a huge artifact, copy yields the capped preview and the "(large output)" branch (which keys on the same 256 KiB) effectively never fires. This is consistent with the expanded-view precedence (`expandedSource` in tool_panel.zig:221), so "copy what's shown" is defensible — flagging only so the large-output criterion isn't assumed to cover artifact tools.

### Verdict
Requesting changes for the two P1s (selection-clobber correctness + CI test coverage). The P2 should be addressed or explicitly scoped. Happy to re-review after.

---REVIEW_POSTED---
internal: counts p0=0 p1=2 p2=1 p3=1

File: zig/src/tui/state.zig (hunk 1 above)
Comment (lsm) [reply zig/src/tui/state.zig]: Fixed. I removed the upsert-time selection mutation and made selectedTool() fall back to the latest tool only when there is no explicit selected_tool_index. I also expanded the state test to cover navigate-to-old-tool followed by a new tool append, verifying the explicit selection is preserved.

File: zig/src/tui/app.zig (hunk 2 above)
Comment (lsm) [reply zig/src/tui/app.zig:1324]: Fixed. flushClipboard now records an error transcript entry when ctx.setClipboard errors or returns false, so clipboard write failures are surfaced instead of silently reporting success.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -47,7 +47,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        group: [core, transport, protocol, providers, utils, makai-cli, agent-types, agent-loop, agent-mod, agent-bridge, agent-unit, agent-chain]
+        group: [core, transport, protocol, providers, utils, makai-cli, tui, agent-types, agent-loop, agent-mod, agent-bridge, agent-unit, agent-chain]
```
Comment (lsm) [inline .github/workflows/ci.yml:50]: **P1 (blocking) — the `tui` CI check added here is RED on Linux.**

Adding the `tui` group is the right fix for round-1 review, but it is the first CI group that compiles `app.zig`'s *test binary* on Linux, and that binary currently fails to **compile**:

```
src/tui/app.zig:32:8: error: dependency on libc must be explicitly specified in the build command
extern "c" fn setenv(...)
referenced by: test.App saves Kimi login credentials as api key: src/tui/app.zig:2927:21
```

Root cause: `zig/src/tui/app.zig:32` declares `extern "c" fn setenv`, which requires libc. On macOS libc is linked implicitly so it builds locally; on Linux CI it must be requested explicitly. `build.zig` already does this for the analogous case — `tui_tests_e2e_mod` sets `.link_libc = true` (build.zig:1100) because `src/tui/tests/e2e_tests.zig:29` declares the same extern — but `tui_app_mod` (build.zig:~1011) is missing it.

(The `makai-cli` group passes because there `app.zig` is only an *imported* module — its test blocks aren't compiled — whereas the `tui` group makes `app.zig` the test root, compiling all its tests including the one referencing `setenv`.)

This is pre-existing on `main` (the extern is in `origin/main`'s app.zig:32 too), so it was just latent until this gate existed. But the PR can't merge with a required check red, and the PR's own 3 app.zig tests don't run in CI until it compiles.

Fix (one line, mirrors the existing pattern): add `.link_libc = true` to `tui_app_mod` in `build.zig`:
```zig
const tui_app_mod = b.createModule(.{
    .root_source_file = b.path("src/tui/app.zig"),
    .target = target,
    .optimize = optimize,
    .link_libc = true,   // <-- add
    .imports = &.{ ... },
});
```
After applying, please re-run CI to confirm the `tui` group is fully green — the build currently aborts at this first error, so any *additional* Linux-only issues in the tui binaries (if any) would only surface once `app.zig` compiles.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (ZhipuAI)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** ZhipuAI

**Recommendation: REQUEST_CHANGES** (own-PR, so COMMENT — GitHub rejects REQUEST_CHANGES from the author. Treat as changes-requested.)

### Round-1 findings — all resolved ✅
Verified each fix in the squashed commit `2bcb067`:
- **P1 selection clobber** — `upsertTool` (state.zig:1401–1417) no longer sets `selected_tool_index`; the "default to latest" moved into `selectedTool()` as a fallback (`self.selected_tool_index orelse self.tools.items.len - 1`, state.zig:671–675). The test now explicitly appends a 3rd tool after navigating to index 0 and asserts the selection stays at 0 — covering the race I flagged. Good.
- **P1 CI tui group** — `tui` added to the matrix (ci.yml:50).
- **P2 clipboard failures** — `flushClipboard` now captures `setClipboard`'s `!bool` return and calls `recordError` on error or `copied == false` (app.zig:1278–1283); `recordError` does `appendTranscript(.@"error", …)` (app.zig:1232–1239), so failures render. Signature confirmed `!bool` in vendor/zigzag context.zig:200.

Local `zig build test-unit-tui` passes (exit 0) and `check-zig-patterns.sh` is clean. Both prior review threads are resolved.

### New P1 (blocking) — the `tui` CI check is RED on Linux
`mergeStateStatus: UNSTABLE` — "Unit Tests - tui" failed. It's a **compile** failure, not a test failure: `app.zig:32` declares `extern "c" fn setenv`, which needs libc. `tui_app_mod` in `build.zig` is missing `.link_libc = true` (the analogous `tui_tests_e2e_mod` already sets it at build.zig:1100 for the same extern). It builds on macOS (implicit libc) but not on Linux CI. This is the first CI group to compile `app.zig`'s test root, so the PR's own 3 app.zig tests currently don't run in CI at all.

Pre-existing on `main` (the extern is in `origin/main`'s app.zig:32 too), surfaced by this PR's new gate — but the PR can't merge with a required check red.

Fix: add `.link_libc = true` to `tui_app_mod` (build.zig:~1011), one line, mirroring the existing pattern. Then re-run CI to confirm the whole `tui` group is green (the build aborts at the first error, so further Linux-only issues, if any, would only appear after this compiles). See anchored comment on ci.yml:50.

### Verdict
One blocking P1 (CI libc). The feature code itself is correct and the round-1 issues are well-addressed — just need the build fix so the CI gate this PR introduces is actually green. Re-review once CI is clean.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (lsm) [reply .github/workflows/ci.yml:50]: Fixed. Added .link_libc = true to tui_app_mod in zig/build.zig, matching the existing e2e module pattern. Re-ran ../scripts/check-zig-patterns.sh, zig build test-unit-tui, and zig build test-unit-makai-cli locally.

File: zig/src/tui/state.zig [hunk 4]
```diff
@@ -668,6 +668,20 @@ pub const AppState = struct {
...
+    pub fn selectedTool(self: *const AppState) ?*const ToolEntry {
+        if (self.tools.items.len == 0) return null;
+        const index = self.selected_tool_index orelse self.tools.items.len - 1;
+        if (index >= self.tools.items.len) return null;
+        return &self.tools.items[index];
+    }
+
+    pub fn selectedToolOutputText(self: *const AppState) ?[]const u8 {
+        const tool = self.selectedTool() orelse return null;
+        if (tool.display_preview.len > 0) return tool.display_preview;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:680]: Acknowledged. This copy path intentionally mirrors the tool panel expanded content precedence in this PR (display_preview, then stored output). Fetching full artifact contents would require adding async artifact retrieval/error handling into the copy path and is beyond this issue slice; large output behavior remains explicit via the preview/artifact hint. Leaving current behavior for now.

File: zig/src/tui/state.zig [hunk 5]
```diff
@@ -668,6 +668,20 @@ pub const AppState = struct {
...
+        if (tool.output.items.len > 0) return tool.output.items;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:681]: Acknowledged. Current ToolEntry.output is the same stored data the tool panel renders today, and this PR copies the selected/focused panel content rather than introducing a new content_json persistence path. Capturing separate tool content text for all local/remote events is a larger state/session schema change and should be handled separately if desired.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by glm-5.1 (ZhipuAI)

> **Model:** glm-5.1 | **Client:** HyperNeo | **Provider:** ZhipuAI

**Recommendation: APPROVE** (own-PR, so COMMENT — GitHub rejects APPROVE from the author. Treat this as an approving verdict.)

### Round 3 — all clear ✅
The libc fix landed exactly right: `.link_libc = true` added to `tui_app_mod` (build.zig:1014), mirroring the existing `tui_tests_e2e_mod` pattern (build.zig:1100) for the same `extern "c" fn setenv` declaration.

Verification:
- **CI "Unit Tests - tui" = SUCCESS** — the gate this PR introduces is now green on Linux. All 13 unit-test groups pass.
- `link_libc` propagation didn't break the `makai-cli` executable (which imports `tui_app_mod`) — "Unit Tests - makai-cli" = SUCCESS.
- Local `zig build test-unit-tui` passes (exit 0); `check-zig-patterns.sh` clean.
- `mergeStateStatus: CLEAN`, `mergeable: MERGEABLE`, all 5 review threads resolved.

### Final state of the feature
- Copy selected/focused tool output via `/copy tool` and `Ctrl+X` (app.zig); both paths flush the clipboard.
- Explicit tool selection is preserved when new tools stream in (no more clobbering) — `upsertTool` no longer mutates `selected_tool_index`; the "default to latest" lives in `selectedTool()` as a null-fallback (state.zig:671–675). The only selection reset is the gated focus-pane entry at state.zig:805. Race covered by test.
- Empty/no-output → transcript message; large output (>256 KiB) → distinct message; clipboard write failures now render an `[error]` transcript entry via `recordError`.
- Tests cover success, empty, large, command dispatch, and selection-tracking (incl. navigate-then-append).

All prior P1/P2 findings from rounds 1 and 2 are resolved with no regressions. Approving — ready for merge to `main` (squash).
